datasetinsights/stats/visualization/keypoints_pose.py

```diff
@@ -0,0 +1,133 @@
+    keypoints = []
+    for ann in annotations:
+        if _is_torso_visible_or_labeled(ann):
+            keypoints.append(ann)
+    return keypoints
+
+
+def _calc_mid(p1: Tuple[Any, Any], p2: Tuple[Any, Any]):
+    """
+    Calculate mid point of two points
+    """
+    return (p1[0] + p2[0]) / 2, (p1[1] + p2[1]) / 2
+
+
+def _calc_dist(p1: Tuple[Any, Any], p2: Tuple[Any, Any]) -> float:
+    """
+    Calculate distance between two points
+    """
+    return math.sqrt(((p1[0] - p2[0]) ** 2) + ((p1[1] - p2[1]) ** 2))
+
+
+def _translate_and_scale_xy(X: List, Y: List):
+    """
+    Return keypoints axis list X and Y after performing translation and scaling.
+    """
+    left_hip, right_hip = (X[11], Y[11]), (X[12], Y[12])
+    left_shoulder, right_shoulder = (X[5], Y[5]), (X[6], Y[6])
+
+    # Translate all points according to mid_hip being at 0,0
+    mid_hip = _calc_mid(right_hip, left_hip)
+    X = [number - mid_hip[0] if number > 0.0 else 0.0 for number in X]
+    Y = [number - mid_hip[1] if number > 0.0 else 0.0 for number in Y]
+
+    # Calculate scale factor
+    scale = (
+        _calc_dist(left_shoulder, left_hip)
+        + _calc_dist(right_shoulder, right_hip)
+    ) / 2
+    X = [number / scale for number in X]
+    Y = [number / scale for number in Y]
+
+    return X, Y
+
+
+def process_annotations(annotations: List) -> Dict:
+    """
+    Process keypoints annotations to extract information for pose plots.
+    Args:
+        annotations (list): List of keypoints lists with format
+        [x1, y1, v1, x2, y2, v2, ...] with the order of COCO_KEYPOINTS
+    Returns:
+        Dict: Processed key-value pair of keypoints name -> (x,y) list.
+    """
+    keypoints = _get_kp_where_torso_visible(annotations)
+
+    processed_kp_dict = {}
+    for name in COCO_KEYPOINTS:
+        processed_kp_dict[name] = {"x": [], "y": []}
+
+    for kp in keypoints:
+        # Separate x and y keypoints
+        x_kp, y_kp = kp[0::3], kp[1::3]
+        x_kp, y_kp = _translate_and_scale_xy(x_kp, y_kp)
+
+        # save keypoints to dict
+        idx = 0
+        for xi, yi in zip(x_kp, y_kp):
+            if xi == 0 and yi == 0:
+                pass
+            elif xi > 2.5 or xi < -2.5 or yi > 2.5 or yi < -2.5:
+                pass
+            else:
+                processed_kp_dict[COCO_KEYPOINTS[idx]]["x"].append(xi)
+                processed_kp_dict[COCO_KEYPOINTS[idx]]["y"].append(yi)
+            idx += 1
+
+    return processed_kp_dict
+
+
+def _get_avg_kp(kp_dict: Dict):
+    """
+    Return average value of keypoints axis list X and Y.
+    """
+    x_avg, y_avg = [], []
+    for key in COCO_KEYPOINTS:
+        kp_x = np.array(kp_dict[key]["x"])
+        kp_y = np.array(kp_dict[key]["y"])
+        x_avg.append(np.mean(kp_x))
+        y_avg.append(np.mean(kp_y))
+    return x_avg, y_avg
+
+
+def get_average_skeleton(kp_dict: Dict, skeleton=COCO_SKELETON) -> List:
+    """
+    return skeleton link list of average keypoints values.
```

Comment:
wdym by link list here?

Reply:
updated doc string of this function: skeleton (a list of connected human joints)